### PR TITLE
#162 Export preferences

### PR DESCRIPTION
**Pull request by mastermichi**
first version

### CODE REVIEW

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 1]
```diff
@@ -396,6 +396,19 @@ public boolean onPreferenceClick(Preference preference) {
                     }
                 });
             }
+
+            Preference exportPreferences = findPreference("exportPreferences");
+            if (exportPreferences != null) {
+                exportPreferences.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
+                        public boolean onPreferenceClick(Preference preference) {
+                        //Toast.makeText(getActivity(), "Export Preference", Toast.LENGTH_LONG).show()
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:404]:
```suggestion
```

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 2]
```diff
@@ -396,6 +396,19 @@ public boolean onPreferenceClick(Preference preference) {
...
+                            ExportPreferences export= new ExportPreferences(getActivity());
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:405]:
```suggestion
                            ExportPreferences export = new ExportPreferences(getActivity());
```

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            share.putExtra(Intent.EXTRA_STREAM, contentUri);
+            context.startActivity(Intent.createChooser(share, "Export"));
+        }
+        catch (Exception ex){
+            Log.e(TAG, "Error sharing file");
+        }
+        return null;
+    }
+
+}
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:114]:
```suggestion
}

```

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+        try {
+            Uri contentUri = FileProvider.getUriForFile(context, "nightdream", newFile);
+            Intent share = new Intent(Intent.ACTION_SEND);
+            share.setType("text/plain");
+            share.putExtra(Intent.EXTRA_STREAM, contentUri);
+            context.startActivity(Intent.createChooser(share, "Export"));
+        }
+        catch (Exception ex){
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:108]: Please restrict to necessary exceptions which can be handled.

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TimeZone;
+
+public class ExportPreferences extends AsyncTask<Void, Void, Void> {
+    public static final String TAG = "ExportPreferences";
+    public static final String PREFS_KEY = "NightDream preferences";
+    private static final String exportFile = "nightdream_export_%s.json";
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:24]:
```suggestion
    private static final String exportFile = "nightclock_preferences_export_%s.json";
```

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    private Context context;
+
+    public ExportPreferences(Context context) {
+        this.context  = context;
+    }
+
+    @Override
+    protected void onPreExecute() {
+        Toast.makeText(context, "Starting Export", Toast.LENGTH_LONG).show();
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:33]:
```suggestion
```

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        if(!exportPath.exists()) {
+            boolean mkdirsResult = exportPath.mkdirs();
+            Log.d(TAG, "Created Directory for export: " + mkdirsResult);
+        }
+
+        // delete old exports
+        File[] oldFiles = exportPath.listFiles();
+        if(oldFiles != null) {
+            for(File oldFile : oldFiles) {
+                if(oldFile.isFile() && oldFile.getName().startsWith("nightdream_export")) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:48]:
```suggestion
                if(oldFile.isFile() && oldFile.getName().startsWith("nightclock_preferences_export_")) {
```

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Override
+    protected void onPreExecute() {
+        Toast.makeText(context, "Starting Export", Toast.LENGTH_LONG).show();
+    }
+
+    @Override
+    protected Void doInBackground(Void... params) {
+        File exportPath = new File(context.getFilesDir(), "export");
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:38]: The export directory should be a static variable. 
private static String exportDir = "export";

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                json.put("Product",      Build.PRODUCT);
+                json.put("Manufacturer", Build.MANUFACTURER);
+                json.put("Model",        Build.MODEL);
+                json.put("Device",       Build.DEVICE);
+                json.put("Display",      Build.DISPLAY);
+                json.put("Timezone",     TimeZone.getDefault().getID());
+                json.put("Offset",       TimeZone.getDefault().getOffset(System.currentTimeMillis()));
+                json.put("ExportTime",   System.currentTimeMillis());
+
+                Map<String, ?> allSP = context.getSharedPreferences(PREFS_KEY, 0).getAll();
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:84]:
```suggestion
                Map<String, ?> allSP = context.getSharedPreferences(Settings.PREFS_KEY, 0).getAll();
```

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+            } catch(JSONException ex) {
+                Log.e(TAG, "Error JSON");
+            }
+
+            outputStream.write(json.toString().getBytes());
+            outputStream.write("\n\n}".getBytes());
+
+            outputStream.close();
+        } catch (Exception ex) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:97]: IOException ?

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+            Log.e(TAG, "Error outputStream");
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:98]: On exception we can leave the entire method and maybe show a toast.

File: src/com/firebirdberlin/nightdream/ExportPreferences.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+import java.io.FileOutputStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TimeZone;
+
+public class ExportPreferences extends AsyncTask<Void, Void, Void> {
+    public static final String TAG = "ExportPreferences";
+    public static final String PREFS_KEY = "NightDream preferences";
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ExportPreferences.java:23]: It's better if we'd refer directly to Settings.PREFS_KEY

```suggestion
```

File: res/xml/filepaths.xml [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<paths xmlns:android="http://schemas.android.com/apk/res/android">
+    <files-path name="export" path="export/"/>
+</paths>
```
Comment (firebirdberlin) [inline res/xml/filepaths.xml:3]:
```suggestion
</paths>

```

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 1 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:404]: done

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 2 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:405]: done

File: src/com/firebirdberlin/nightdream/ExportPreferences.java (hunk 3 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ExportPreferences.java:114]: done

File: src/com/firebirdberlin/nightdream/ExportPreferences.java (hunk 5 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ExportPreferences.java:24]: done

File: src/com/firebirdberlin/nightdream/ExportPreferences.java (hunk 9 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ExportPreferences.java:84]: done

File: src/com/firebirdberlin/nightdream/ExportPreferences.java (hunk 12 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ExportPreferences.java:23]: done

**Pull request merged**